lib/minfraud/transaction.rb

```diff
@@ -28,60 +32,139 @@ class Transaction
     attr_accessor :avs_result, :cvv_result
 
     # Miscellaneous attribute (optional)
-    attr_accessor :requested_type, :forwarded_ip
-
+    attr_accessor :requested_type
+
+    ATTRIBUTES = [
+      :ip, :txn_id, :license_key,
+      :city, :state, :postal, :country,
+      :ship_addr, :ship_city, :ship_state, :ship_postal, :ship_country,
+      :email_domain, :email_md5, :phone,
+      :bin,
+      :user_agent, :accept_language,
+      :amount, :currency, :txn_type, :shop_id,
+      :avs_result, :cvv_result,
+      :requested_type
+    ].freeze
+
+    INPUT_STRING_ATTRIBUTES = [
+      :ip, :txn_id, :license_key,
+      :city, :state, :postal, :country,
+      :ship_addr, :ship_city, :ship_state, :ship_postal, :ship_country,
+      :email, :phone,
+      :bin,
+      :user_agent, :accept_language,
+      :currency, :txn_type,
+      :avs_result, :cvv_result,
+      :requested_type
+    ].freeze
+
+    INPUT_NUMERIC_ATTRIBUTES = [:amount].freeze
+
+    REQUIRED_ATTRIBUTES = [:ip, :txn_id, :license_key].freeze
+
+    CONVERT_STRING_ATTRIBUTES = [:shop_id].freeze
+
+    # Initializes the Transaction using parameters set from the block
+    # @raise [TransactionError] if parameters are set incorrectly
     def initialize
       yield self
       unless has_required_attributes?
         raise TransactionError, 'You did not set all the required transaction attributes.'
       end
       validate_attributes
+      freeze
     end
 
     # Hash of attributes that have been set
     # @return [Hash] present attributes
     def attributes
-      attrs = [:ip, :city, :state, :postal, :country, :license_key, :ship_addr, :ship_city, :ship_state, :ship_postal, :ship_country, :email_domain, :email_md5, :phone, :bin, :session_id, :user_agent, :accept_language, :txn_id, :amount, :currency, :txn_type, :avs_result, :cvv_result, :requested_type, :forwarded_ip, :shop_id]
-      attrs.map! do |a|
-        [a, send(a)]
-      end
-      Hash[attrs]
+      Hash[ATTRIBUTES.map { |a| [a, send(a)] }].compact
     end
 
     private
 
     # Ensures the required attributes are present
     # @return [Boolean]
     def has_required_attributes?
-      ip && txn_id && license_key
+      REQUIRED_ATTRIBUTES.none? { |attr| send(attr).nil? || send(attr).empty? }
```

Comment:
[`Object#blank?`](https://apidock.com/rails/Object/blank%3F) might be able to help here 🙂

Reply:
apparently #blank? and #present? are Rails methods, so they can't be used here 😅

Reply:
Ahh of course 🤦